qiskit/visualization/counts_visualization.py

```diff
@@ -127,6 +128,143 @@ def plot_histogram(
             display(hist1, hist2)
 
     """
+    if not isinstance(data, list):
+        data = [data]
+    for dat in data:
+        if isinstance(dat, (QuasiDistribution, ProbDistribution)):
+            raise VisualizationError(
+                "Input must be Counts or a dict, " "consider using 'plot_distribution'"
+            )
+    return _plotting_core(
+        data,
+        figsize,
+        color,
+        number_to_keep,
+        sort,
+        target_string,
+        legend,
+        bar_labels,
+        title,
+        ax,
+        filename,
+        kind="counts",
+    )
```

Comment:
Instead of raising here we should instead keep the current behavior for this release (and potentially emit a `PendingDeprecationWarning` warning). Then we can raise a `DeprecationWarning` in 0.23.0 when a distribution object is passed in and then raise in 0.25.0.  The problem with raising like this is despite the function name being a a mismatch for the functionality for < 0.22.0 this was the only function people could use to plot a distribution objects previously and this will actively break their code when they upgrade to 0.22.0. Instead we should do something like:

```suggestion
    for dat in data:
        kind = "counts"
        if isinstance(dat, (QuasiDistribution, ProbDistribution)) or isinstance(next(data.values()), float):
            kind = "distribution"

    return _plotting_core(
        data,
        figsize,
        color,
        number_to_keep,
        sort,
        target_string,
        legend,
        bar_labels,
        title,
        ax,
        filename,
        kind=kind,
    )
```

Reply:
Yeah so this is where some issues get raised.  So if it were just plan probability distributions then one could go down this route.  However, the quasi-probabilities can go negative.  So if we allow for plotting them here, then either we plot incorrectly as the lower bound is zero, or we extent the lower bound to be negative (which `plot_distribution` does) but then further mess up the notion of plotting a "histogram".  I guess one could check the values and see if any are negative and then raise only on that case, but I am sure that that would be breaking as well, but perhaps for a smaller subset of people.

Reply:
`ProbDistribution` I think we agree can pass through regardless.

Reply:
Ok, so I have modified this so that `plot_histogram` allows everything through and changes the y-axis is any input is a distribution.  This will in turn trigger `kind=distribution` which will modify the bounds to be negative if needed.

The downside of this is that by further manipulating `plot_histogram` we just enforce it as the tool for plotting all distributions, and at that point why bother adding `plot_distribution` at all?  We could provide a warning that hints the user to move to `plot_distribution` without breaking their code.

Reply:
Yeah I think using a `PendingDeprecationWarning` here would make sense. We can promote it to a full `DeprecationWarning` in 0.23.0 so it'll get in users faces more

Reply:
I added a warning in: https://github.com/Qiskit/qiskit-terra/pull/8762/commits/ad78b75776a151380487a1c9c45e0248bebbe769